### PR TITLE
feat: Add Exception Handling for LiteLLM

### PR DESCRIPTION
**Pull request by dyga01**
1. **Pull Request:** Add Exception Handling for API key and API server exceptions for LiteLLM so it doesn't crash and give a long stack trace.

2. Aidan Dyga (@dyga01) and Hemani Alaparthi (@hemanialaparthi)

3. #3 

4. bug, enhancement

5. This pull request aims to enhance LiteLLM's error handling capabilities, focusing on API key and server exceptions. By implementing the new exceptions.py file and making changes to advise.py, we hope to prevent crashes and long stack traces that can confuse users. The goal is to improve the library's stability and user experience by providing more graceful error handling and user-friendly error messages, making LiteLLM more robust and easier to debug in real-world scenarios.

6. Coverage will be maintained for the exceptions.py file, as we have implemented new tests to ensure the new feature is fully tested.

7. The tests have been conducted on a Mac. It would be great if I could have a Linux and Windows user test it out.

8. Output can vary based on the exceptions.

i) For example for the command where the `advice_model` is incorrect, the output should be the according:

```cmd
Exception Type: NotFoundError
Explanation: The requested resource was not found. Please check if your model or endpoint is correct.

If your issue persists, ensure the model you entered is listed below:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free
- openrouter/google/gemma-2-9b-it:free
```

ii) If the `advice_server` is invalid, this should be the output:

```cmd
Exception Type: APIConnectionError
Explanation: There was a connection issue to the server.
NOTE: This error can sometimes be caused by an invalid server URL. Please verify the URL you're using.

If your issue persists, ensure the model you entered is listed below:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free
- openrouter/google/gemma-2-9b-it:free
```

iii) NOTE: If LiteLLM does change their exception handling, the program will _still not_ crash because we implemented a default behaviour for non LiteLLM Exceptions and this will produce the error message and a general purpose output but not a stacktrace .

### CODE REVIEW

**Comment by gkapfham** [comment]
Hello @dyga01 and @hemanialaparthi there are two PRs connected to exception handling. Do you have a preference on the order in which they are reviewed and ultimately merged?

**Comment by dyga01** [comment]
> Hello @dyga01 and @hemanialaparthi there are two PRs connected to exception handling. Do you have a preference on the order in which they are reviewed and ultimately merged?

We do not have a preference for the order in which both PRs are merged. We would suggest merging them chronologically, but we do not foresee any issues.

**Review by PCain02 — CHANGES_REQUESTED** [review]
Some notes from testing on Windows:

So I changed the advice model and got the intended output: 
image
I noticed that it still graded my work however the coding mentor did not work (expected) 

However, when I change the advice server it gives me just `The --advice-server option did not specify a valid URL` and it does not grade my work. I do not think this was the intended output because the idea is that it would still grade like usual and then when it came to the mentor section create an error. 

image

The commands I ran were `poetry run execexam . ./tests/test_question_one.py --report trace --report status --report failure --report code 
--report setup --advice-method apiserver --advice-model pickles --advice-server https://execexamadviser.fly.dev/ --report advice --fancy --debug`  (works as intended) 
and `poetry run execexam . ./tests/test_question_one.py --report trace --report status --report failure --report code 
--report setup --advice-method apiserver --advice-model anthropic/claude-3-haiku-20240307 --advice-server rotten --report advice --fancy --debug` (errored before any grading)

I would suggest a change be made so that the typical grading code still can be executed but when it gets to the LLM part it throws an error so both exceptions with the advice model and the advice server work the same way.

**Comment by dyga01** [comment]
@PCain02 and @hemanialaparthi - I just made changes to the code to remove the sys.exit statements. I believe these were causing it to exit early. Please retry the commands Pallas specified to ensure this works as intended. Thanks!

**Comment by hemanialaparthi** [comment]
> @PCain02 and @hemanialaparthi - I just made changes to the code to remove the sys.exit statements. I believe these were causing it to exit early. Please retry the commands Pallas specified to ensure this works as intended. Thanks!

Thank you for working on this! I tested this on a Mac by running the command with an invalid advice server url and I got the intended output which includes the failing checks!

```cmd

Exception Type: APIConnectionError
Explanation: There was a connection issue to the server.
NOTE: This error can sometimes be caused by an invalid server URL. Please verify the URL you're using.

If your issue persists, ensure the model you entered is listed below:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free
- openrouter/google/gemma-2-9b-it:free

╭─────── Overall Status ───────╮
│                              │
│ ✘ One or more checks failed. │
│                              │
╰──────────────────────────────╯
```

**Review by PCain02 — APPROVED** [review]
Good! You fixed the issue with not grading when the advice server is incorrect. I would say have it print suggestions for advice-servers but apparently that is not okay because people should have their own and it would cost Professor a lot of money to provide a server for everyone.

**Comment by hannahb09** [comment]
I ran the exact same two commands as Pallas and I got the right outputs. I am on a linux. Also I am not set as a reviewer but I do approve this pr.

**Comment by hemanialaparthi** [comment]
> I ran the exact same two commands as Pallas and I got the right outputs. I am on a linux. Also I am not set as a reviewer but I do approve this pr.

Just added you as a reviewer! Sorry for that.

**Comment by hemanialaparthi** [comment]
> Good! You fixed the issue with not grading when the advice server is incorrect. I would say have it print suggestions for advice-servers but apparently that is not okay because people should have their own and it would cost Professor a lot of money to provide a server for everyone.

Hi @dyga01 and @PCain02. I discussed this with @gkapfham and he said yes, we can't provide them with the server URLs as it _might_ dent his credit card (my words, not his) but he advised us to direct students to the LiteLLM Proxy Server link where they can get more information about server URLs. The latest push to this branch add this feature! Would be amazing if you guys could test it and let me know if the intended output is shown:

```cmd
Exception Type: APIConnectionError
Explanation: There was a connection issue to the server.
NOTE: This error can sometimes be caused by an invalid server URL. Please verify the URL you're using.

If your issue persists, ensure the model you entered is listed below:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free

For server connectivity issues, please visit https://docs.litellm.ai/docs/simple_proxy for a valid LiteLLM proxy.
```

Thanks!

**Review by CalebKendra — APPROVED** [review]
Exception handling works well, consider adding these messages to a `--fancy` outputs block format. potentially in the `overall_status` block

**Review by boulais01 — APPROVED** [review]
LGTM

**Comment by gkapfham** [comment]
Hello @hemanialaparthi, what do you think about @CalebKendra's comment about the debugging output being governed by the use of the `--fancy` flag. That seems to make sense to me, but it would be greatly appreciated if you can explore that further. Thanks!

**Comment by dyga01** [comment]
> Hello @hemanialaparthi, what do you think about @CalebKendra's comment about the debugging output being governed by the use of the `--fancy` flag. That seems to make sense to me, but it would be greatly appreciated if you can explore that further. Thanks!

@gkapfham and @CalebKendra Adding this feature would definitely be helpful, but we do not want to interfere with the changes Molly will be making to the UI. Overall, we think we should integrate this after this PR in an upcoming PR about the UI.

**Comment by hemanialaparthi** [comment]
> > Hello @hemanialaparthi, what do you think about @CalebKendra's comment about the debugging output being governed by the use of the `--fancy` flag. That seems to make sense to me, but it would be greatly appreciated if you can explore that further. Thanks!
> 
> @gkapfham and @CalebKendra Adding this feature would definitely be helpful, but we do not want to interfere with the changes Molly will be making to the UI. Overall, we think we should integrate this after this PR in an upcoming PR about the UI.

Hi @gkapfham! @dyga01 and I fixed the merge issues and this PR is ready to be merged after being tested!

**Comment by gkapfham** [comment]
Hello @hemanialaparthi and @dyga01 this PR needs a review from one additional student who was not a part of the team who implemented the feature. (You have one extra review from a TL, but you still need one more review from a student in our software engineering team).

**Review by hannahb09 — APPROVED** [review]
I checked this pr before when i was not an approver but since I am now I officially approve it .

**Comment by gkapfham** [comment]
Hello @hemanialaparthi and @dyga01 why does the output from @rebekahrudd show a specific `new_suggested_url` two times? Why is this output even needed?

**Comment by hemanialaparthi** [comment]
> Hello @hemanialaparthi and @dyga01 why does the output from @rebekahrudd show a specific `new_suggested_url` two times? Why is this output even needed?


I have an answer for this. @rebekahrudd didn't want to share the name of the models here for safety reasons so she just changed the names of the models. But, I was with here when she ran it and it had the intended output.

**Comment by rebekahrudd** [comment]
Here is a screenshot of my output
output

**Comment by hemanialaparthi** [comment]
Hello @gkapfham! I've worked on improving the output for the Exception Handling PR to enhance clarity and accuracy. Now, when an invalid advice model/server is requested, the output will guide users more effectively with specific suggestions for model selection and links to valid LiteLLM models and proxies.

```cmd
Exception Type: NotFoundError
Explanation: The requested resource was not found. Please check if your model and/or endpoint is correct.

If your issue persists, ensure the model you entered is correct, such as:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free

Please visit https://docs.litellm.ai/docs/providers for more valid LiteLLM models

For server connectivity issues, please visit https://docs.litellm.ai/docs/simple_proxy for a valid LiteLLM proxy.
```

**Comment by gkapfham** [comment]
Is it possible to give more details than NotFoundError?

**Comment by rebekahrudd** [comment]
I believe the explanation bellow the exception type is there to tell more about the error. The exception type and explanation are saved and a dictionary inside the `exceptions.py` file. Does the explanation need to be improved?

**Comment by gkapfham** [comment]
Hello @rebekahrudd and @hemanialaparthi, the prior output shows examples like this:

```
Exception Type: APIConnectionError
```

My remaining concern is that using `APIConnectionError` does not indicate that this is an error specifically coming from the LLM interaction. While I agree that this is the title of the PR and thus it is an issue that is apparent to all of us who are working on the tool, I am trying to empathize with a person who is using the tool and then see this error and does not understand that the specific problem is an API that supports LLM communication. As such, I am suggesting that your further clarify the error message. See what I mean? What do you think?

**Comment by hemanialaparthi** [comment]
> Hello @rebekahrudd and @hemanialaparthi, the prior output shows examples like this:
> 
> ```
> Exception Type: APIConnectionError
> ```
> 
> My remaining concern is that using `APIConnectionError` does not indicate that this is an error specifically coming from the LLM interaction. While I agree that this is the title of the PR and thus it is an issue that is apparent to all of us who are working on the tool, I am trying to empathize with a person who is using the tool and then see this error and does not understand that the specific problem is an API that supports LLM communication. As such, I am suggesting that your further clarify the error message. See what I mean? What do you think?

@gkapfham, Thanks! I understand what you mean. I have been working on making the exception messages more detailed and I want to know what you think about these messages:

```py
litellm_exceptions = {
    "NotFoundError": "The requested LLM resource was not found. Please check if your model and/or endpoint is correct.",
    "AuthenticationError": "There was an issue with your authentication while accessing the LLM API. Please verify your API key.",
    "RateLimitError": "You've hit the rate limit for LLM API usage. Please try again later or adjust your usage.\nNOTE: This error can sometimes be caused by an invalid API key.",
    "InvalidRequestError": "Your request to the LLM API was malformed. Please check the parameters you've sent.",
    "APIError": "An internal LLM API error occurred. Please try again later.",
    "APIConnectionError": "There was a connection issue with the LLM server.\nNOTE: This error can sometimes be caused by an invalid server URL. Please verify the URL you're using.",
}
```

Would you want me to add more details and description or does this suffice.

**Comment by gkapfham** [comment]
Hello @hemanialaparthi can you please consider making the descriptions of the errors shorter and more succinct? Can you also please show what this would look like when run in the terminal window?

**Comment by hemanialaparthi** [comment]
> Hello @hemanialaparthi can you please consider making the descriptions of the errors shorter and more succinct? Can you also please show what this would look like when run in the terminal window?

Yes! These are the new outputs, I have come up with that are both short and concise.

This is how it handles invalid API Models:

```cmd
Exception Type: NotFoundError
Explanation: LLM resource not found. Please check your model and/or endpoint.

If your issue persists, ensure the model you entered is correct, such as:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free

Please visit https://docs.litellm.ai/docs/providers for more valid LiteLLM models

For server connectivity issues, please visit https://docs.litellm.ai/docs/simple_proxy for a valid LiteLLM proxy.
```

This is how it handles invalid servers:

```cmd
Exception Type: APIConnectionError
Explanation: Connection failed. 
NOTE: This error can sometimes be caused by an invalid server URL. Verify your server URL.

If your issue persists, ensure the model you entered is correct, such as:
- anthropic/claude-3-haiku-20240307
- anthropic/claude-3-opus-20240229
- groq/llama3-8b-8192
- openrouter/meta-llama/llama-3.1-8b-instruct:free

Please visit https://docs.litellm.ai/docs/providers for more valid LiteLLM models

For server connectivity issues, please visit https://docs.litellm.ai/docs/simple_proxy for a valid LiteLLM proxy.
```

And, here are the other descriptions of the errors:

```python
    litellm_exceptions = {
        "NotFoundError": "LLM resource not found. Please check your model and/or endpoint.",
        "AuthenticationError": "API authentication failed. Please verify your API key.",
        "RateLimitError": "Rate limit exceeded. Wait and retry or check API key.\nNOTE: This error can sometimes be caused by an invalid API key.",
        "InvalidRequestError": "Malformed API request. Please review parameters.",
        "APIError": "Internal LLM API error. Retry later.",
        "APIConnectionError": "Connection failed. \nNOTE: This error can sometimes be caused by an invalid server URL. Verify your server URL.",
    }
```

**Review by gkapfham — COMMENTED** [review]
Thanks for all of your work on this feature @dyga01 and @hemanialaparthi. This looks like a useful feature. Before we merge it can you please confirm that this should _only change the behavior of `execexam` when it encounters an exception with the LLM_? Is that correct?

**Comment by dyga01** [comment]
> Thanks for all of your work on this feature @dyga01 and @hemanialaparthi. This looks like a useful feature. Before we merge it can you please confirm that this should _only change the behavior of `execexam` when it encounters an exception with the LLM_? Is that correct?

@gkapfham That is correct. This feature is specifically for LiteLLM exceptions.

**Comment by CalebKendra** [comment]
We should consider adding `dismiss stale pull request approvals` as a branch protection rule if we want to change PR's after approval.

**Comment by dyga01** [comment]
> For PRs, if a feature is still in development, it should be labeled as a draft, not an active PR. Can we confirm that no further development is needed? @hemanialaparthi @dyga01

@CalebKendra This PR is complete and ready to be merged.

**Review by CalebKendra — APPROVED** [review]
LGTM. Thanks for working hard on this one, looks good.

**Review by boulais01 — APPROVED** [review]
LGTM

File: execexam/advise.py
Comment (PCain02) [inline execexam/advise.py:84]: So just curious why did you remove the return codes? Where they causing errors? I also noticed `sys.exit` is no longer in use why is that the case? I am asking because a lot of my work has been in `advise.py` and I will also have to add these changes to mine.

**Review by PCain02 — APPROVED** [review]
I ran the code on Windows 10 in various ways and it provided the correct output 
image

I also ran the new tests and they all passed. 

I made some comments/questions on the code changes because I would like to knowledge share about the `advise.py` file. 

Overall, looks very good.

File: execexam/advise.py
Comment (hemanialaparthi) [reply execexam/advise.py:84]: We removed the exit codes because @gkapfham emphasized that there should be only one exit point for the program, which should be in `main.py`. This design choice improves debugging as it centralizes the program’s termination. By removing sys.exit from advise.py, we ensure the program only exits through main.py, making it easier to track where and why the program terminates.

Comment (PCain02) [reply execexam/advise.py:84]: Okay that makes a lot of sense thank you!

**Pull request merged**
